### ISSUE TITLE
Add configuration of files to ignore

### ISSUE THREAD

**Issue by Dbz** [adrienverge/yamllint#43]
A great feature for this project would be the ability to exclude certain files. This way, if you are adding yamllint to a project, you can ignore every file with issues until you get around to fixing them. Right now, the only way to achieve this is through comments on every file.

A good way to accomplish this is in the rules configuration:

``` YAML
rules:
  indentation:
    indent-sequences: false
    spaces: 2
    exclude:
    - 'configs/config_yaml.yml'
```

A second add-on feature would be to allow these excludes to be auto-generated. This second feature helps make it easy to integrate yamllint into an existing codebase.

Right now our project has 11400 errors, so these features would make yamllint an idea yaml linter for our project, but as it stands, there is too much noise to lint new yamls as they come.

**Comment by adrienverge** [comment]
Hi!

Looks like a duplicate of #5.

> What you describe looks very useful. However I'm not sure such filtering should be part of yamllint's logic, especially because it would be redundant with common UNIX tools.
>
> For instance, what about using find?
>
> ```shell
> yamllint $(find foo -name '*.yml' ! -name 'b*.yml' ! -name 'd.yml')
> ```

**Comment by Dbz** [comment]
Hey Adrien,

This is not a duplicate of #5. We want to exclude files from specific rules and not the linter in general, so this is not something that can be done with UNIX tools. I feel very strongly that this is a very important feature for a production yaml linter, and I want to give you insight to our pipeline:

1. Devs need to be able to run the linter easily.

We have ten dev teams around the company committing to this repository. They can be given the line (or script) like `yamllint $(find foo -name '*.yml' ! -name 'b*.yml' ! -name 'd.yml')` with fifty more file names included to run, but it causes some issues. For example, the is no up-to-date list of the work that needs to be done to clean these files up, and there is also no priority on the syntax vs. style issues. We want linting to be very easy, and we want refactoring to be easy.

2. We have indentation issues that are causing teams who work on the same yaml many problems. 

We want to use a linter to fix the most important problems in the yamls first. Inconsistencies are causing issues for many teams. We render partial yamls inside yamls, and indentation/duplicate keys, etc. are causing problems and need to be dealt with immediately. If we have to ignore entire files, then we may miss these issues as they pop up.

3. The CI pipeline needs to run the linter, and that requires constantly updating the yamllint run command.

The command excludes entire files, so it actually can not be used as part of the CI pipeline. The CI needs to lint all of the yamls (unless there is a very specific reason why not to, and style issues are not a good reason).

I totally understand and appreciate your resistance to duplicating system functionality - it goes well with the philosophy of _create tools that do one thing and one thing well_. However, there is no system functionality that can replicate this feature, and it's a core feature for a production-level linter (#5 gave examples of a couple other linters that include the features). I also think #5 should included as a feature. It's technically already a feature through commenting, but this is more structured and allows devs to go back and see what work needs to be done to clean up the project:

``` YAML
rules:
  all:
    exclude:
    - 'configs/config_yaml.yml'
```

My coworker and I want to integrate this linter into our pipeline, but we're having issues running and debugging the project. However, we believe we can achieve excluding files from rules in `YamlLintConfig#enabled_rules` in the `config.py` file (and we also need to parse that exclude list from the config). Would you be willing to implement this feature or help us develop it in a fork of the project? We are big supporters of Open Source, and Yamllint would be used by a medium/big company in production! We think that would be pretty cool!

A feature after this would be to auto-generate the list of files to exclude (so that this is an easy linter to add to an existing project). However, one feature at a time!

Thank you so much for considering this proposal. Even if this isn't something you want to add, **we really appreciate you taking the time to hear us out**.

**Comment by adrienverge** [comment]
Hi @Dbz,

Thanks, I better understand your goal.

I don't like adding include / exclude paths *per rule*, perhaps doing the opposite (divide your project into *path patterns* and then apply rules to these divisions) would be more readable?

For example, I'm thinking about this kind of config format:

```yaml
- files: |
    *.yaml                     # .gitignore-style pattern matching,
    !*.dontlint.yaml           # e.g. using pathspec lib
    !permissive-dir/*.yaml
    !very-strict-dir/*.yaml
  extends: default

- files: |
    permissive-dir/*.yaml
    !*.dontlint.yaml
  extends: default
  rules:
    line-length:
      max: 100

- files: |
    very-strict-dir/*.yaml
    !*.dontlint.yaml
  extends: default
  rules:
    document-start:
      level: error
    truthy:
      level: error
    indentation:
      spaces: 4
```

What do you think?

**Comment by graaff** [comment]
For reference the pattern that @Dbz is describing is similar to the one used by https://github.com/bbatsov/rubocop 

rubocop has a nice option --auto-gen-config which generates a config file that excludes a current linter errors into a separate todo configuration file. It does that by listing the specific files as exclude options for each individual rule (or disabled the rule if there are too many files with issues). This means that you can add a passing CI check to a legacy project and start making incremental improvements.

What this manner of specification does for me is to say that, yes, I want this rule to apply, and these files still need to be fixed. The rule does apply to any new files out of the box, helping to improve the codebase one step at a time. This approach lets me focus on gradually applying the rules that I want to all files.

Your prososed syntax on the other hand tells me that not all YAML files in my project come equal, and there are different rule sets for each one. Now I'm thinking in terms of sets of files that my project consists of and possible set of rules to apply. I can see how that could be useful when different sets of YAML files really do need different static rules, but I don't think this syntax would help me focus on improving my code base.

Obviously the same end goal could be achieved with both approaches, but I feel that the config proposed by @Dbz would be helping me achieve my goals better.

**Comment by hassek** [comment]
Any updates? I would love this feature included!

**Comment by adrienverge** [comment]
Hi, to the best of my knowledge no one is working on this. Again, any help to find the best way to implement it, then do it, is welcome.

**Comment by Dbz** [comment]
Hey @adrienverge, let's try and hammer out a design so that it is ready to be implemented!

I recommend going with a style similar to the one I provided for the reasons @graaff presented as it makes it very easy to incrementally clean up the codebase. I do think it may be worth having two separate files- one for rules, and one for exclusions similar to how RuboCop does it:

yamllint.yml:
```yaml
extends: default

rules:
  indentation:
    indent-sequences: false
    spaces: 2
```
yamllint-todo.yml:
```yaml
indentation:
    exclude:
    - 'configs/config_yaml.yml'
```

This is a much cleaner approach as it separates the purposes of the files, and sets us up to run a `yamllint -c yamllint.yml --auto-gen-config` which will generate `yamllint-todo.yml`.

**Comment by hassek** [comment]
Excluding files from a specific check (like indentation in the example) is quite peculiar, to have a general exclude would meet almost all cases and should be simpler to implement.

```yaml
rules:
  exclude:
    - "filename.yml"
    - "*.json"
```

Excluding specifics would be handy too, so my proposal would be to implement the general exclude and the specifics exclude in separate branches.

**Comment by Dbz** [comment]
Hey @hassek, specifically excluding rules on a per-file basis is something other linters such as robocop provide.

Indentation is a great case study because values are often templated into yamls (and yamls are templated into yamls). Incorrect indentation will invalidate the rendered yaml, so while you want to exclude your file from the other rules, you can't exclude it because you need to check the indentation rule. However, if files are excluded on a rule by rule basis, it is very easy to add this to an existing code base and clean it up incrementally based on what is most important (indentation, then duplicate keys, etc.).

**Comment by hassek** [comment]
@Dbz _rubocop_ does a great job at that, it will be useful to have the same approach here. Let's make this happen. Just thinking out loud about the complexities of the tasks and what can we do first and later.

If I am not mistaken this would be the requirements to implement:

* Global exclusion of files
* Per rule exclusion of files
* _yamllint-todo_ generator

We can separate these 3 tasks and they will be all be useful by themselves.

Please let me know if I am missing some point.

**Comment by Dbz** [comment]
That looks good to me 👍

**Comment by adrienverge** [comment]
Hi @Dbz, @hassek,

I'm not fond of using another config file if there is no reason to it.

What about using @Dbz's initial spec, but add to it a global ignore pattern?
Also, it would be great to use the commonly-used and powerful `.gitingore` format, using pathspec lib. For example:

```yaml
---

# This is a global ignore list, for files on which the linter shouldn't be run
# at all. It will ignore files ending in `.dontlint.yaml`, files in the `dist`
# directory, and files in `bin`, except when they end in `.lint-me-anyway.yaml`.
ignore: |
  *.dontlint.yaml
  /bin/
  !/bin/*.lint-me-anyway.yaml
  /dist/

extends: default

rules:
  indentation:
    indent-sequences: false
    spaces: 2
    # This is an ignore list for this rule only.
    ignore: |
      /generated/
      *.template.yaml
```

This way, `yamllint -c config.yaml --auto-gen-config` could generate a file based on `config.yaml`, with per-rule `ignore` sections added.

What do you think?

**Comment by adrienverge** [comment]
I have implemented a prototype of my previous proposition on the [`feat/ignore-paths` branch](https://github.com/adrienverge/yamllint/tree/feat/ignore-paths).

Can you test it and tell me if it suits your needs?

```shell
pip3 install --user pathspec
git clone https://github.com/adrienverge/yamllint.git -b feat/ignore-paths
cd yamllint
python3 -m yamllint -c your-config.yaml your/directory
```

**Comment by hassek** [comment]
@adrienverge I'll try it out on my project. Following the `.gitignore` format it's 🥇

**Comment by ssbarnea** [comment]
Any chance of seeing this fixed? Using `.gitignore` seems like a nice-and-clean solution to me.

I really hate getting linting errors from inside the `.tox/` folder, which is already ignored.

**Comment by adrienverge** [comment]
@ssbarnea feel free to test the [`feat/ignore-paths` branch](https://github.com/adrienverge/yamllint/tree/feat/ignore-paths) and tell us what you think!

Until I have more feedback, you can still use `find` to exclude your `.tox` folder:
```shell
yamllint $(find .. -name '*.yaml' -not -path './tox/*')
```

**Comment by ssbarnea** [comment]
I was able to workaround by limiting the search to a specifict folder, luckly I got all files under one folder.

I still hope to see the .gitignore implemented. It is much better than duplicating ignore logic in another place.

**Comment by bensymonds** [comment]
For what it's worth, I've given this a go and it worked for me! My case was very simple though - just excluding a directory:

```
ignore:
  /vendor/
```

Looking forward to seeing it released.

**Comment by Dbz** [comment]
Me too! Great teamwork by everyone involved to thoroughly discuss and plan this 👍

**Comment by adrienverge** [comment]
I'm glad you like it!

yamllint 1.8.0 was just released, and here is the related documentation: https://yamllint.readthedocs.io/en/latest/configuration.html#ignoring-paths

**Comment by bensymonds** [comment]
Quick work, thanks 👍 I'll update my thing to use the released version and shout if there are any problems.

**Comment by hassek** [comment]
Awesome! cheers!
